runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/StreamingStepMetricsContainer.java

```diff
@@ -178,4 +192,109 @@ public static Iterable<CounterUpdate> extractMetricUpdates(
+      if (Duration.between(entry.getValue(), extractionTime)
+              .compareTo(maximumPerWorkerCounterStaleness)
+          > 0) {
+        RemoveSafeDeltaCounterCell cell =
+            new RemoveSafeDeltaCounterCell(entry.getKey(), perWorkerCounters);
+        cell.deleteIfZero();
+        deletedMetricNames.add(entry.getKey());
+      }
+    }
+
+    // Insert new zero-valued metrics into `perWorkerCountersByFirstStaleTime`.
+    currentZeroValuedCounters.forEach(
+        name -> perWorkerCountersByFirstStaleTime.putIfAbsent(name, extractionTime));
+
+    // Metrics in 'deletedMetricNames' have either been removed from 'perWorkerCounters' or are no
+    // longer zero-valued.
+    perWorkerCountersByFirstStaleTime.keySet().removeAll(deletedMetricNames);
+  }
+
+  /**
+   * Extracts metric updates for all PerWorker metrics that have changed in this Container since the
+   * last time this function was called. Additionally, deletes any PerWorker counters that have been
+   * zero valued for more than {@code maximumPerWorkerCounterStaleness}.
+   */
+  private Iterable<PerStepNamespaceMetrics> extractPerWorkerMetricUpdates() {
+    ConcurrentHashMap<MetricName, Long> counters = new ConcurrentHashMap<MetricName, Long>();
+    ConcurrentHashMap<MetricName, HistogramData> histograms =
+        new ConcurrentHashMap<MetricName, HistogramData>();
+    HashSet<MetricName> currentZeroValuedCounters = new HashSet<MetricName>();
+
+    // Extract metrics updates.
+    perWorkerCounters.forEach(
+        (k, v) -> {
+          Long val = v.getAndSet(0);
+          if (val == 0) {
+            currentZeroValuedCounters.add(k);
+            return;
+          }
+          counters.put(k, val);
+        });
+    perWorkerHistograms.forEach(
+        (k, v) -> {
+          HistogramData val = v.getCumulative().getAndReset();
+          if (val.getTotalCount() == 0) {
+            return;
+          }
+          histograms.put(k.getKey(), val);
+        });
+
+    deleteStaleCounters(currentZeroValuedCounters, Instant.now(clock));
+
+    return MetricsToPerStepNamespaceMetricsConverter.convert(stepName, counters, histograms);
+  }
+
+  /**
+   * @param metricsContainerRegistry Metrics will be extracted for all containers in this registry.
+   * @return An iterable of {@link PerStepNamespaceMetrics} representing the changes to all
+   *     PerWorkerMetrics that have changed since the last time this function was invoked.
+   */
+  public static Iterable<PerStepNamespaceMetrics> extractPerWorkerMetricUpdates(
+      MetricsContainerRegistry<StreamingStepMetricsContainer> metricsContainerRegistry) {
+    return metricsContainerRegistry
+        .getContainers()
+        .transformAndConcat(StreamingStepMetricsContainer::extractPerWorkerMetricUpdates);
+  }
+
+  // The following methods are used to test that zero-valued perWorkerMetrics metrics are properly
+  // deleted.
+  @VisibleForTesting
+  Map<MetricName, AtomicLong> getPerWorkerCounters() {
```

Comment:
Is it intended that these are able to mutated by callers? If not recommended to use ImmutableMap https://engdoc.corp.google.com/eng/doc/devguide/java/practices/choosing-collections.md?cl=head#immutable

Reply:
Removed this function.